File: artifacts/embedding_worker.py
import time
import os
from typing import List, Dict
from pinecone_client import PineconeClient

# Mock Embedding Function (Replace with OpenAI call)
def generate_embeddings(texts: List[str]) -> List[List[float]]:
    # Returns random vectors for demo
    import numpy as np
    return [np.random.rand(1536).tolist() for _ in texts]

class EmbeddingWorker:
    def __init__(self):
        self.pc_client = PineconeClient()
        self.batch_size = 10

    def process_job_batch(self, jobs: List[Dict]):
        """
        jobs: List of dicts with {'id': str, 'text': str, 'metadata': dict}
        """
        print(f"🚀 Processing batch of {len(jobs)} embedding jobs...")
        
        # 1. Generate Embeddings (Batch)
        texts = [j['text'] for j in jobs]
        embeddings = generate_embeddings(texts)
        
        # 2. Prepare Vectors for Pinecone
        vectors = []
        for job, vec in zip(jobs, embeddings):
            vectors.append((job['id'], vec, job['metadata']))
            
        # 3. Upsert to Pinecone
        self.pc_client.upsert_batch(vectors)
        
        # 4. (Mock) Update Mongo Status
        print(f"✅ Successfully upserted {len(vectors)} vectors. Updating Mongo status...")

if __name__ == "__main__":
    # Simulate a queue of jobs
    mock_jobs = [
        {"id": "c_101", "text": "@tech_daily Daily Tech News", "metadata": {"handle": "@tech_daily"}},
        {"id": "c_102", "text": "@art_studio Digital Art Tips", "metadata": {"handle": "@art_studio"}},
        {"id": "c_103", "text": "@fitness_pro Workout Routines", "metadata": {"handle": "@fitness_pro"}},
    ]
    
    worker = EmbeddingWorker()
    worker.process_job_batch(mock_jobs)

File: src/embeddingWorker.ts
import { PineconeClient } from './pineconeClient';

// Mock Embedding Function (Replace with OpenAI call)
function generateEmbeddings(texts: string[]): number[][] {
    // Returns random vectors for demo
    return texts.map(() => Array.from({ length: 1536 }, () => Math.random()));
}

interface Job {
    id: string;
    text: string;
    metadata: {
        handle: string;
        [key: string]: any;
    };
}

export class EmbeddingWorker {
    private pcClient: PineconeClient;
    private batchSize: number;

    constructor() {
        this.pcClient = new PineconeClient();
        this.batchSize = 10;
    }

    async processJobBatch(jobs: Job[]) {
        console.log(`🚀 Processing batch of ${jobs.length} embedding jobs...`);

        // 1. Generate Embeddings (Batch)
        const texts = jobs.map(j => j.text);
        const embeddings = generateEmbeddings(texts);

        // 2. Prepare Vectors for Pinecone
        const vectors = jobs.map((job, index) => ({
            id: job.id,
            values: embeddings[index],
            metadata: job.metadata
        }));

        // 3. Upsert to Pinecone
        await this.pcClient.upsertBatch(vectors);

        // 4. (Mock) Update Mongo Status
        console.log(`✅ Successfully upserted ${vectors.length} vectors. Updating Mongo status...`);
    }
}

if (require.main === module) {
    (async () => {
        // Simulate a queue of jobs
        const mockJobs: Job[] = [
            { id: 'c_101', text: '@tech_daily Daily Tech News', metadata: { handle: '@tech_daily' } },
            { id: 'c_102', text: '@art_studio Digital Art Tips', metadata: { handle: '@art_studio' } },
            { id: 'c_103', text: '@fitness_pro Workout Routines', metadata: { handle: '@fitness_pro' } },
        ];

        const worker = new EmbeddingWorker();
        await worker.processJobBatch(mockJobs);
    })();
}
